File: C#/KnightProbabilityInChessboard.cs
public class Solution {
    double[,,] knight;
    int[][] moves = new int[][] {
        new int[] {-2, -1},
        new int[] {-2, 1},
        new int[] {2, -1},
        new int[] {2, 1},
        new int[] {-1, -2},
        new int[] {1, -2},
        new int[] {-1, 2},
        new int[] {1, 2}
    };
    
    public double KnightProbability(int n, int k, int row, int column) {
        knight = new double[n, n, k + 1];
        for(int i = 0; i < n; i++){
            for(int j = 0; j < n; j++){
                for(int z = 0; z <= k; z++){
                    knight[i, j, z] = -1.0;
                }
            }
        }
        return solve(n, k, row, column);
    }
    
    public double solve(int n, int k, int row, int column) {
        if(row < 0 || row >= n || column < 0 || column >= n){
            return 0.0;
        }
        if(k == 0){
            return 1.0;
        }
        if(knight[row, column, k] != -1.0){
            return knight[row, column, k];
        }
        double prob = 0.0;
        foreach(int[] move in moves){
            prob += solve(n, k - 1, row + move[0], column + move[1]) / 8.0;
        }
        knight[row, column, k] = prob;
        return prob;
    }
}

File: Java/Knight Probability In Chessboard/KnightProbabilityInChessboard.java
class Solution {
    public double knightProbability(int n, int k, int row, int column) {
        double knight[][][] = new double[n][n][k+1];
        // Possible Moves
        int[][] moves = {{-2, -1}, {-2, 1}, {2, -1}, {2, 1}, {-1, -2}, {1, -2}, {-1, 2}, {1, 2}};

        for(int i = 0; i < n; i++) {
            for(int j = 0; j < n; j++) {
                knight[i][j][0] = 1.0;
            }
        }

        // z is moves left
        for(int z = 1; z <= k; z++){
            for(int i = 0; i < n; i++){
                for(int j = 0; j < n; j++){
                    for(int[] move : moves){
                        int newI = i + move[0];
                        int newJ = j + move[1];
                        if(newI >= 0 && newI < n && newJ >= 0 && newJ < n){
                            // We have to sum up the probabilities
                            knight[i][j][z] += knight[newI][newJ][z - 1] / 8.0;
                        }
                    }
                }
            }
        }
        return knight[row][column][k];
    }
}